### PR TITLE
Add support for WVFPDP

### PR DESCRIPTION
**Pull request by hakonhagland**
Implements support for keyword WVFPDP. See also companion PR https://github.com/OPM/opm-simulators/pull/4620 for more information.

### CODE REVIEW

**Comment by hakonhagland** [comment]
jenkins build this opm-simulators=4620 please

**Comment by hakonhagland** [comment]
jenkins build this opm-simulators=4620 please

**Comment by hakonhagland** [comment]
Jenkins then reports another error of the same type:
```
/var/lib/jenkins/workspace/opm-common-PR-builder/mpi/install/include/opm/common/utility/Serializer.hpp:551:46: error: invalid use of incomplete type ‘using T1 = using element_type = using element_type = std::remove_extent<Opm::WVFPDP>::type’ {aka ‘class Opm::WVFPDP’}
  551 |             const_cast<PtrType&>(data).reset(new T1);
      |                                              ^~~~~~
In file included from /var/lib/jenkins/workspace/opm-common-PR-builder/deps/opm-simulators/tests/test_ParallelSerialization.cpp:99:
/var/lib/jenkins/workspace/opm-common-PR-builder/mpi/install/include/opm/input/eclipse/Schedule/Well/Well.hpp:69:7: note: forward declaration of ‘using T1 = using element_type = using element_type = std::remove_extent<Opm::WVFPDP>::type’ {aka ‘class Opm::WVFPDP’}
   69 | class WVFPDP;
      |       ^~~~~~
```
This one should be fixed by https://github.com/OPM/opm-simulators/pull/4620/commits/f98927688b19951ade5d3ab1a03af177b995ff4a

**Comment by hakonhagland** [comment]
jenkins build this opm-simulators=4620 please

**Comment by hakonhagland** [comment]
Next jenkins error was reported like this:
```
[NON-XML-CHAR-0x1B][1;31;49m*** 1 failure is detected in the test module "Glift1"
[NON-XML-CHAR-0x1B][0;39;49munknown location(0): [NON-XML-CHAR-0x1B][4;31;49mfatal error: in "G1": memory access violation at address: 0x00000000: no mapping at fault address[NON-XML-CHAR-0x1B][0;39;49m
```
I believe this was a segfault due to a null pointer was returned at line 195: https://github.com/OPM/opm-simulators/blob/61e5ce3d731fd6cd2473359aa87561bd7232a38c/tests/test_glift1.cpp#L195. I think happened because the WVFPDP object was not initialized correctly and it resulted in a failed BHP from THP computation.

**Comment by hakonhagland** [comment]
jenkins build this opm-simulators=4620 please

**Review by bska — COMMENTED** [review]
Thanks a lot for the contribution.  I don't think this is quite ready for merging yet.  I'd like a set of unit tests that demonstrate the component working correctly, and how call sites are supposed to use the `getDP()` function.  As written, it's not completely clear to me how the function is expected to work.

I'm also a little concerned about the naming of the member functions.

Finally, I think we might need to turn at least the scaling factor data member into a `std::optional<double>` instead in order to track whether or not the item and/or keyword has been defaulted.  We need that piece of informatoin when creating our restart files.

**Comment by hakonhagland** [comment]
@bska Thanks for the review! 

> I'd like a set of unit tests that demonstrate the component working correctly

Could these tests be added to this file?

https://github.com/OPM/opm-common/blob/master/tests/test_rst.cpp

File: src/opm/output/eclipse/AggregateWellData.cpp [hunk 1]
```diff
@@ -1027,6 +1028,20 @@ namespace {
...
+                                    SWProp&&         swprop,
+                                    SWellArray&      sWell)
+        {
+            using Ix = VI::SWell::index;
+            using M = ::Opm::UnitSystem::measure;
+
+            const auto& options = well.getWVFPDP();
+
+            sWell[Ix::VfpBhpAdjustment] = swprop(M::pressure, options.getDPDP());
+            sWell[Ix::VfpBhpScalingFact] = options.getDPFP();
```
[the remark this answers is not in the document]
Comment (hakonhagland) [reply src/opm/output/eclipse/AggregateWellData.cpp:1042]: > the scaling factor has the sentinel value -1.0e+20

@bska Interesting. Why do these values have these particular values? If they are unused, we could pick any value couldn't we? For example, why is item 90 defaulted to 1.0 whereas item 91 is defaulted to `dflt` ?

[the remark this answers is not in the document]
Comment (bska) [reply src/opm/output/eclipse/AggregateWellData.cpp:1042]: >  Why do these items have these particular values?

Compatibility reasons.  We want to enable restarting other simulators based on simulation results from Flow, and that put constraints on what values we can choose here.

Tangentially, how did you determine that items 90 and 91 were the right ones for this keyword?

[the remark this answers is not in the document]
Comment (hakonhagland) [reply src/opm/output/eclipse/AggregateWellData.cpp:1042]: > Compatibility reasons

@bska I see. Good to know!

> how did you determine that items 90 and 91 were the right ones for this keyword?

I assumed that we could choose any unused items in the SWell array? Maybe that is not the case then? How can we determine which items in the SWell array that can be used for new keywords?

[the remark this answers is not in the document]
Comment (bska) [reply src/opm/output/eclipse/AggregateWellData.cpp:1042]: >> how did you determine that items 90 and 91 were the right ones for this keyword?
>
> I assumed that we could choose any unused items in the SWell array? Maybe that is not the case then?

No, that is not the case.  We need to be rather careful about what entries we put into `SWEL` and in what locations.

>  How can we determine which items in the SWell array that can be used for new keywords?

I'll tell you offline.

[the remark this answers is not in the document]
Comment (hakonhagland) [reply src/opm/output/eclipse/AggregateWellData.cpp:1042]: > I'll tell you offline.

A short update after a discussion with @bska and @tskille offline: The items should be written to indices 78 and 83 in the SWell array. These indices have correct default values (0.0 and 1.0), see https://github.com/OPM/opm-common/blob/b38cedf11346d722c9df589400a643c0c8998e54/src/opm/output/eclipse/AggregateWellData.cpp#L657

**Comment by hakonhagland** [comment]
jenkins build this opm-simulators=4620 please

**Comment by bska** [comment]
jenkins build this please

**Comment by hakonhagland** [comment]
> jenkins build this please

@bska Why did you run the jenkins build without  #4620 in opm-simulators?

**Comment by bska** [comment]
> > Because I wanted to see if this PR could be merged independently of that one.
> 
> Ok I see. Seems like it cannot be merged independently then?

That's my assessment too and it makes bringing this in a little, though not a lot, more fiddly.

**Comment by hakonhagland** [comment]
jenkins build this opm-simulators=4620 please

**Comment by hakonhagland** [comment]
jenkins build this opm-simulators=4620 please

**Comment by hakonhagland** [comment]
jenkins build this opm-simulators=4620 please

**Comment by tskille** [comment]
The latest commits has been tested and results looks very good. Looking forward to see this go into master.

**Comment by totto82** [comment]
@hakonhagland Can you fix the new issues. There is some unused variables.

**Comment by hakonhagland** [comment]
jenkins build this opm-simulators=4620 please

**Comment by hakonhagland** [comment]
Hmm.. not sure why the jenkins build now fails with https://ci.opm-project.org/job/opm-common-PR-builder/5821/console : 

```
/var/lib/jenkins/workspace/opm-common-PR-builder/deps/opm-simulators/ebos/eclgenericvanguard.cc:313:29:   required from here
/var/lib/jenkins/workspace/opm-common-PR-builder/mpi/install/include/opm/common/utility/Serializer.hpp:551:46: error: invalid use of incomplete type ‘using T1 = using element_type = using element_type = std::remove_extent<Opm::WVFPDP>::type’ {aka ‘class Opm::WVFPDP’}
  551 |             const_cast<PtrType&>(data).reset(new T1);
      |                                              ^~~~~~
In file included from /var/lib/jenkins/workspace/opm-common-PR-builder/deps/opm-simulators/ebos/eclgenericvanguard.cc:55:
/var/lib/jenkins/workspace/opm-common-PR-builder/mpi/install/include/opm/input/eclipse/Schedule/Well/Well.hpp:69:7: note: forward declaration of ‘using T1 = using element_type = using element_type = std::remove_extent<Opm::WVFPDP>::type’ {aka ‘class Opm::WVFPDP’}
   69 | class WVFPDP;
      |       ^~~~~~
```
The missing header file `WVFPDP.hpp` was added to `eclgenericvanguard.cc` in a previous commit, so I do not understand why it now suddenly complains about this again..

**Comment by bska** [comment]
> The missing header file `WVFPDP.hpp` was added to `eclgenericvanguard.cc` in a previous commit, so I do not understand why it now suddenly complains about this again.

I've seen this kind of symptom&mdash;previous build failures that come back later&mdash;when GitHub for some reason gets confused about its branches and their branch points.  The first course of action in this case is always to rebase all branches on top of the current master sources.  Doing so **typically** solves the issue, but in the case that it doesn't the next step is typically to temporarily close the PRs for a while (15 - 30 minutes is my rule of thumb).

**Comment by hakonhagland** [comment]
jenkins build this opm-simulators=4620 please

**Comment by hakonhagland** [comment]
> The first course of action in this case is always to rebase all branches on top of the current master sources. 

@bska Thanks for the insight, I have rebased this PR on the current master. Let's see if it helps.

**Comment by hakonhagland** [comment]
> I have rebased this PR on the current master. Let's see if it helps.

@bska Unfortunately, rebasing did not work. Do you see any reason for the error message that I might have overlooked?

**Comment by bska** [comment]
> > have rebased this PR on the current master. Let's see if it helps.
> 
> Unfortunately, rebasing did not work.

Did you really rebase the branch onto the latest master version?  This PR page looks a little different than I would normally expect if you did do the rebase.  As far as I can tell you've rebased onto master as of PR #3549, but there have been more commits since then.  You also need to rebase OPM/opm-simulators#4620 onto the very latest version of OPM-Simulators.

> Do you see any reason for the error message that I might have overlooked?

No, not really.  I'd suggest rebasing the branches again and retrying the build.  If it still does not go through we'll have to look at other measures.

**Comment by blattms** [comment]
We have been struggelling with these kinds of error a lot lately. Usually the problem is this:

At the place where the error occurs the type is indeed not fully defined (the header has not been fully read), just the forward declaration has been seen. Usually this is due to some cyclic in inclusion of the header. The best way to cope is usually to find the function triggering the problem (calling the serialize method), move that function definition (and maybe others) out of the class declaration. Then move problematic header includes from above the  class declaration to after it and add needed forward declarations above.

**Comment by blattms** [comment]
On the other hand the new header is really innocent.
Seems like serializeOp in the Well (and hence in Schedule) needs the definition. Could be as easy as including the new header at the top of Well.hpp.
Otherwise try the approach pointed out above and move the searializeOp method out of class declaration

**Comment by hakonhagland** [comment]
jenkins build this opm-simulators=4620 please

**Comment by totto82** [comment]
Thanks for the work. Merging...

**Pull request merged**